internal/dnfjson/dnfjson.go

```diff
@@ -0,0 +1,414 @@
+			if len(transactions[dsIdx].RepoIDs) < len(prevRepoIDs) {
+				return nil, nil, fmt.Errorf("chained packageSet %d does not use all of the repos used by its predecessor", dsIdx)
+			}
+
+			for idx, repoID := range prevRepoIDs {
+				if repoID != transactions[dsIdx].RepoIDs[idx] {
+					return nil, nil, fmt.Errorf("chained packageSet %d does not use all of the repos used by its predecessor", dsIdx)
+				}
+			}
+		}
+	}
+
+	dnfRepoMap, err := s.reposFromRPMMD(repos)
+	if err != nil {
+		return nil, nil, err
+	}
+	args := arguments{
+		Repos:        dnfRepoMap,
+		Transactions: transactions,
+	}
+
+	req := Request{
+		Command:          "depsolve",
+		ModulePlatformID: s.modulePlatformID,
+		Arch:             s.arch,
+		CacheDir:         s.cacheDir,
+		Arguments:        args,
+	}
+
+	return &req, rpmRepoMap, nil
+}
+
+// Helper function for creating a dump request payload
+func (s *Solver) makeDumpRequest(repos []rpmmd.RepoConfig) (*Request, error) {
+	dnfRepos, err := s.reposFromRPMMD(repos)
+	if err != nil {
+		return nil, err
+	}
+	req := Request{
+		Command:          "dump",
+		ModulePlatformID: s.modulePlatformID,
+		Arch:             s.arch,
+		CacheDir:         s.cacheDir,
+		Arguments: arguments{
+			Repos: dnfRepos,
+		},
+	}
+	return &req, nil
+}
+
+// convert internal a list of PackageSpecs to the rpmmd equivalent and attach
+// key and subscription information based on the repository configs.
+func (pkgs packageSpecs) toRPMMD(repos map[string]rpmmd.RepoConfig) []rpmmd.PackageSpec {
+	rpmDependencies := make([]rpmmd.PackageSpec, len(pkgs))
+	for i, dep := range pkgs {
+		repo, ok := repos[dep.RepoID]
+		if !ok {
+			panic("dependency repo ID not found in repositories")
+		}
+		dep := pkgs[i]
+		rpmDependencies[i].Name = dep.Name
+		rpmDependencies[i].Epoch = dep.Epoch
+		rpmDependencies[i].Version = dep.Version
+		rpmDependencies[i].Release = dep.Release
+		rpmDependencies[i].Arch = dep.Arch
+		rpmDependencies[i].RemoteLocation = dep.RemoteLocation
+		rpmDependencies[i].Checksum = dep.Checksum
+		rpmDependencies[i].CheckGPG = repo.CheckGPG
+		if repo.RHSM {
+			rpmDependencies[i].Secrets = "org.osbuild.rhsm"
+		}
+	}
+	return rpmDependencies
+}
+
+// Request command and arguments for dnf-json
+type Request struct {
+	// Command should be either "depsolve" or "dump"
+	Command string `json:"command"`
+
+	// Platform ID, e.g., "platform:el8"
+	ModulePlatformID string `json:"module_platform_id"`
+
+	// System architecture
+	Arch string `json:"arch"`
+
+	// Cache directory for the DNF metadata
+	CacheDir string `json:"cachedir"`
+
+	// Arguments for the action defined by Command
+	Arguments arguments `json:"arguments"`
+}
+
+// arguments for a dnf-json request
+type arguments struct {
+	// Repositories to use for depsolving
+	Repos []repoConfig `json:"repos"`
+
+	// Depsolve package sets and repository mappings for this request
+	Transactions []depsolveArgs `json:"transactions"`
+}
+
+type depsolveArgs struct {
```

Comment:
super-nitpick: since you agreed to rename this to `Transactions`, it may make sense to rename also the `depsolveArgs` type to something like `transactionArgs`. Totally nice to have though...